### ISSUE TITLE
Slows to a crawl after about 36,000 inserts under luajit 2.1.0-alpha

### ISSUE THREAD

**Issue by sharpobject** [grandquista/Lua-ReQL#7]
I'm not sure what the cause is here. I was able to insert all the data I wanted to anyway by inserting more than 1 row at a time, but I'd like to figure out what's causing the slowdown.

![screen shot 2015-10-27 at 4 06 20 am](https://cloud.githubusercontent.com/assets/579947/10756682/f05c4b24-7c62-11e5-9fb0-5932d50c0f0e.png)

**Comment by grandquista** [comment]
If the query is the same 36k times then I'm not sure where the code path would change. Guessing at things that might affect this...

- There are some issues with the server being synchronous that might be fixed by upgrading the protocol version on the connection. I haven't gotten my tests to pass with that change yet.
- Testing with an `r.pool` instead of `r.connect` would be another idea.
- If there are functions in your query then the anonymous function builder would have a large global integer incrementing for variable names to send to the server.
- If the query is not being sent with `no_reply = true` then there is a lot of data possibly on the wire.

**Comment by sharpobject** [comment]
The original thing was reading from some huge file, so that would not make a very good example. I was able to get the same behavior with the following program:
```local r = require"rethinkdb"
local json = require"dkjson"
local socket = require("socket")

local c, err = r.connect()
if err then error(err) end
c:use("test")
r.table_create("junk"):run(c)

for id=1,500000 do
  print(id)
  r.table("junk"):insert({{id=id}}):run(c)
end
c:close()```

The script stopped at 41365. The data explorer says the biggest thing I've stored has ID 37567.

I'm definitely hitting your last case of not sending ```no_reply = true``` but this behavior is still surprising to me. While writing this comment the script was able to submit 4 more items.

**Comment by grandquista** [comment]
Making a test from that example now.

**Comment by grandquista** [comment]
I can definitely reproduce this locally. Looking into what might be causing it.
Canceling the script when it pauses always throws from `socket:send` making this look like a network saturation issue.
The insert line can be changed to `r.table("junk"):insert({id=id}):run(c)` removing the need to pack and unpack an array on both sides of the network. This got me from pausing at 19975 to pausing at 25865 in my tests.

**Comment by grandquista** [comment]
Leaving the script to see how it completes. I will have time to look at this again later this evening.

**Comment by grandquista** [comment]
I'm almost certain this is caused by the server timing out and calls to send silently failing after the default 20 second period. That also explains the data explorer returning a lower id than the script says was last sent. Inserting 5 arrays of 100000 elements each like the test documents doesn't take much time.

Time to read about luasocket and see if I can look for a closed socket better.

**Comment by sharpobject** [comment]
I looked into this a little. The most recent Lua-ReQL makes that script eventually crash, but it still claims to have inserted more things than it really did.

I added logging to all the ```raw_socket:receive()``` and ```raw_socket_send()``` call sites. This script never runs ```raw_socket:receive()``` other than when connecting. It seems like it should? Or maybe I need to supply a callback to get it to try reading replies?

[Someone on the internet] (http://stackoverflow.com/questions/16054394/how-to-check-a-socket-is-closed-or-not-in-luasocket-library/16074095#16074095) said in luasocket you had to try reading from a socket to be sure whether it was closed, but I'm not sure how much data you have to try reading.

**Comment by sharpobject** [comment]
This script terminates successfully

```local r = require"rethinkdb"
local json = require"dkjson"
local socket = require("socket")

local c, err = r.connect()
if err then error(err) end
c:use("test")
r.table_create("junk"):run(c)

for id=1,500000 do
  print(id)
  r.table("junk"):insert({id=id}):run(c, function() end)
  if id % 100 == 0 then
    c.raw_socket:settimeout(0)
    c.raw_socket:receive("*a")
    c.raw_socket:settimeout(c.timeout)
  end
end
c:close()```

So I guess the earlier one fails because we never read the replies.

**Comment by grandquista** [comment]
The original script should run all queries in no reply mode since it has no callback. I would expect nothing to ever be sent from the server.

If a cursor is being read from then it will check for the closed socket as SO shows.

I made a patch for send to error based on the luasocket documentation. That wasn't passing the test suite at the end of last weekend.

**Comment by sharpobject** [comment]
If I remove the callback, the server still sends data. Here's some of it:

```
{"t":2,"r":[],"n":[]}{"t":2,"r":[],"n":[]}{"t":2,"r":[],"n":[]}{"t":2,"r":[],"n":[]}{"t":2,"r":[],"n":[]}{"t":2,"r":[],"n":[]}{"t":2,"r":[],"n":[]}{"t":2,"r":[],"n":[]}
```

**Comment by grandquista** [comment]
That is definitely a bug in my driver not sending `noreply` to the server correctly.

**Comment by sharpobject** [comment]
I think it sends ```noreply``` correctly but there's some heartbeat/ping thing that is sent by the server regardless, and we fail to read that.

**Comment by sharpobject** [comment]
I mean, the stuff the server sent in the case without callbacks was all like ```{"t":2,"r":[],"n":[]}``` but with callbacks we also got things that look like replies, ```{"t":1,"r":[{"deleted":0,"errors":1,"first_error":"Duplic...```

**Comment by grandquista** [comment]
Looking at https://github.com/rethinkdb/rethinkdb/blob/next/drivers/python/rethinkdb/net.py#L436 there is no cursor and no response read from the server for `noreply`. Thats how my driver treats it.

`{"t":2,"r":[],"n":[]}` is an empty array success, but none of these queries look like they return arrays. It is also missing the header giving a token and the length of the response.

With a callback my driver expects the cursor passed to handle requesting data from the server. Those responses after the first look correct and have extra non character data that I presume decodes to the token and size expected.

In short your second test today is doing what I would expect. The first script is getting back data that the official drivers aren't even handling.

**Comment by grandquista** [comment]
Can I get your RethinkDB version to test the python driver for these responses?

**Comment by sharpobject** [comment]
rethinkdb 2.1.5+2~0precise (GCC 4.6.3)

I can install some other version if you want!

**Comment by grandquista** [comment]
After wrapping the python driver socket I'm not seeing any unexpected responses like `{"t":2,"r":[],"n":[]}`. I'm back to looking at my driver for a bug in the query serialization.

**Comment by grandquista** [comment]
I will release v0.7.0 tonight and it will error on your original test. Finding the issue with unexpected data on the line will still keep this open.

**Comment by grandquista** [comment]
The latest commit has this test passing without erroring, and confirms the last id can be read from the server. The socket and protocol code has been broken out and changed to a source/sink model.
